website/docs/configuration/expressions/references.html.md

```diff
@@ -60,24 +60,58 @@ For more information about how to use resource references, see
 
 `var.<NAME>` is the value of the [input variable](/docs/configuration/variables.html) of the given name.
 
+If the variable has a type constraint (`type` argument) as part of its
+declaration, Terraform will automatically convert the caller's given value
+to conform to the type constraint.
+
+For that reason, you can safely assume that a reference using `var.` will
+always produce a value that conforms to the type constraint, even if the caller
+provided a value of a different type that was automatically converted.
+
+In particular, note that if you define a variable as being of an object type
+with particular attributes then only _those specific attributes_ will be
+available in expressions elsewhere in the module, even if the caller actually
+passed in a value with additional attributes. You must define in the type
+constraint all of the attributes you intend to use elsewhere in your module.
+
 ### Local Values
 
 `local.<NAME>` is the value of the [local value](/docs/configuration/locals.html) of the given name.
 
+Local values can refer to other local values, even within the same `locals`
+block, as long as you don't introduce circular dependencies.
+
 ### Child Module Outputs
 
-* `module.<MODULE NAME>.<OUTPUT NAME>` is the value of the specified
-  [output value](/docs/configuration/outputs.html) from a
-  [child module](/docs/configuration/blocks/modules/index.html) called by the
-  current module.
+`module.<MODULE NAME>` is an value representing the results of
```

Comment:
I did consider adding more elaborate examples to all of the sections on this page, but this was the last thing I was working on so I ran out of time and so went with just revising the existing text a little for now. I agree that it would be nice to expand each of these sections with a full configuration example (not just an expression) which shows both the declaration of a thing _and_ a separate reference to that thing, to try to connect the declaration syntax with the reference syntax.